Feature/initial commit

linda-mccall opened the pull request:
db schema, entity relationship models, service classes, REST controllers, JWT authentication

linda-mccall committed:
initial commit

linda-mccall committed:
fix issue with holiday foreign keys

Commit:
adding enums and test cases

linda-mccall committed:
fix issue with json errors and composite keys

paul-doherty2018 committed:
committing unit tests

paul-doherty2018 committed:
commitng change to employee service and adding unit tests

linda-mccall committed:
refactor

linda-mccall committed:
Merge branch 'feature/initial-commit' of https://github.com/UnosquareBelfast/AdminCore into feature/initial-commit

# Conflicts:
#	src/main/java/com/unosquare/adminCore/service/EmployeeService.java

linda-mccall committed:
add additional search functionality, finish writing unit tests.

remove unused swagger-login page

linda-mccall committed:
add functionality for teamLeader, systemAdmin and user roles

linda-mccall committed:
implement multi-stage docker file

linda-mccall committed:
implement multi-stage docker file,

update holidays so that you can have half days,
new unit ,
write script to generate db schema,
add logic to update create

BrownMark commented on pom.xml:
```diff
@@ -0,0 +1,129 @@
+        </dependency>
+
+        <dependency>
+            <groupId>com.microsoft.sqlserver</groupId>
```
Is this required?

BrownMark commented on pom.xml:
```diff
@@ -0,0 +1,129 @@
+    <packaging>jar</packaging>
+
+    <name>hr-manager-service</name>
+    <description>Demo project for Spring Boot</description>
```
Can we keep it to admin core for now? Just to fall in line with the rest of unosquare's internal conventions

BrownMark commented on src/main/java/com/unosquare/adminCore/ApplicationStartup.java:
```diff
@@ -0,0 +1,45 @@
+        employeeUserRoleService.save(EmployeeUserRoles.User.getRole());
+
+        if (employeeService.findByEmail("superuser@unosquare.com") == null) {
+            SignUpRequest signUpRequest = new SignUpRequest("super", "user",
```
This is fine for now, we'll be switching over to an SSO provider in the future anyway

BrownMark commented on src/main/java/com/unosquare/adminCore/HrManagerServiceApplication.java:
```diff
@@ -0,0 +1,23 @@
+
+@SpringBootApplication
+@EnableSwagger2
+public class HrManagerServiceApplication extends SpringBootServletInitializer {
```
Can we rename this to something inline with admin core?

BrownMark commented on src/main/java/com/unosquare/adminCore/entity/Holiday.java:
```diff
@@ -0,0 +1,59 @@
+        this.endDate = endDate;
+        this.holidayLength = holidayLength;
+        this.employee = employee;
+        this.status = "new";
```
Suspect a copy paste error here.

BrownMark commented on src/main/java/com/unosquare/adminCore/enums/EmployeeUserRoles.java:
```diff
@@ -0,0 +1,28 @@
+
+import com.unosquare.adminCore.entity.EmployeeUserRole;
+
+public enum EmployeeUserRoles {
```
Can we refactor to EmployeeRoles, with Employee representing the user in the system I think this will cause confusion.

BrownMark commented:
We need to add the following types as data models for referential integrity:

1. Country
2. Client Status
3. Contract Status
4. Holiday Status
5. Security roles

The enums should be updated with the IDs for these fields (we'll need a migration strategy for this  (e.g. https://flywaydb.org/getstarted/firststeps/maven)- you'll need to set the IDs explicitly, this would be done with set identity_insert on sql server and is a very manual process with postgres).

BrownMark commented:
We need to get away from the dal data models being passed in the spring controllers. It ties our hands, can result in lazy loading serialization problems and can result in a lot of meaningless data being sent across the wire.

We should be creating DTOs for now, i'd take a look at model mapper or similar to save some time: http://modelmapper.org/getting-started/

BrownMark commented on Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM openjdk:8
```
Openjdk is pretty large, this definitely lends itself well to a multi stage dockerfile

BrownMark commented on src/main/java/com/unosquare/adminCore/ApplicationStartup.java:
```diff
@@ -0,0 +1,45 @@
+    EmployeeService employeeService;
+
+    @Autowired
+    EmployeeUserRoleService employeeUserRoleService;
```
Yeah lets change this to EmployeeRole, it sounds like it's joining employee, user and roles

BrownMark commented on src/main/java/com/unosquare/adminCore/entity/Contract.java:
```diff
@@ -0,0 +1,38 @@
+
+    }
+
+    public Contract(ContractPK id, String status) {
```
the status needs to be an integer here with a separate table for referential integrity.

BrownMark commented on src/main/java/com/unosquare/adminCore/entity/Holiday.java:
```diff
@@ -0,0 +1,59 @@
+        this.endDate = endDate;
+        this.holidayLength = holidayLength;
+        this.employee = employee;
+        this.status = "new";
```
I don't think the holiday status enum has a value of "new", again there should really be another table for the ref data.

BrownMark commented on src/main/java/com/unosquare/adminCore/controller/ContractController.java:
```diff
@@ -0,0 +1,64 @@
+public class ContractController {
+
+    @Autowired
+    ContractService contractService;
```
I'll add a PR wide comment, but we really shouldn't be serializing data models for our request. Separate out into DTOs for requests to keep the payload to what you need and reduce the risk of lazy loading (here's a quick overview https://howtoprogramwithjava.com/hibernate-eager-vs-lazy-fetch-type/ )

BrownMark commented on src/main/java/com/unosquare/adminCore/entity/Holiday.java:
```diff
@@ -0,0 +1,59 @@
+    private LocalDate startDate;
+    private LocalDate endDate;
+
+    private short holidayLength;
```
Is holiday length needed? Having something like this places a requirement on having to keep it up to date and as we add more criteria in it will require tasks to keep in current.

I'd keep it to the date range only for now, performance wouldn't be an issue.

BrownMark reviewed (changes requested):
Hey Linda, some changes, nothing too major though. Largely looks great

trojanspike committed:
docker run env setup

linda-mccall committed:
implement multi-stage docker file,
update holidays so that you can have half days,
new unit tests,
write script to generate db schema,
update statuses etc to use Ids,
map clients and contracts to DTOs in rest controllers

linda-mccall committed:
Merge remote-tracking branch 'origin/feature/initial-commit' into feature/initial-commit

# Conflicts:
#	Dockerfile

trojanspike committed:
Change docker image used , add CrossOrigin and allow js api hits

linda-mccall committed:
DTO mapping in git requests, add cross origin mappings, add http option mappings

linda-mccall committed:
implement flyway db migration,
update schema database script,
remove unused imports in project

BrownMark commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM maven:3.5-jdk-8 as BUILD
+ 
+COPY src /usr/src/myapp/src
+COPY src/main/resources/application.properties.docker /usr/src/myapp/src/main/resources/application.properties
```
can we update the target directory to not contain myapp.

BrownMark commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+ 
+FROM openjdk:8-jre-alpine
+ 
+COPY --from=BUILD /usr/src/myapp/target/admin-core-0.0.1-SNAPSHOT.jar /opt/jboss/wildfly/standalone/deployments/admin-core.jar
```
Can we change the target directory to not include jboss wildfly etc, this was copied and pasted from a sample

BrownMark commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+    <parent>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.0.0.RELEASE</version>
```
Is it worth using 2.1 here? I seem to remember seeing news on flaws present in 2.0

BrownMark commented on src/main/java/com/unosquare/adminCore/controller/ClientController.java:
```diff
@@ -0,0 +1,88 @@
+    @Autowired
+    ClientService clientService;
+
+    private ModelMapper modelMapper = new ModelMapper();
```
We should probably inject this. Gives us flexibility in controlling scope etc

BrownMark commented on src/main/java/com/unosquare/adminCore/controller/ClientController.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private List<ClientDto> mapClientsToDtos(List<Client> clients){
+        return clients.stream().map(client -> modelMapper.map(client, ClientDto.class)).collect(Collectors.toList());
```
I'm pretty sure model mapper supports list mapping e.g. modelMapper.map(clients,  ArrayList<Client>.class)

BrownMark reviewed (changes requested):
Just some minor final changes

BrownMark commented:
Last note really, if you get sonar lint installed can you run a quick scan in the app and fix any errors/warnings

trojanspike committed:
Add docker files, docker proxy and notes folder added

trojanspike committed:
Merge branch 'feature/initial-commit' of github.com:UnosquareBelfast/AdminCore into feature/initial-commit

trojanspike committed:
build container with Makefile - edit notes

trojanspike committed:
add docker-compose file to run the containers

trojanspike committed:
make docker files into docker folder. edit docker readme files

linda-mccall committed:
code cleanup

linda-mccall committed:
Merge remote-tracking branch 'origin/feature/initial-commit' into feature/initial-commit

trojanspike committed:
add data folder mount to persist data

trojanspike committed:
Merge branch 'feature/initial-commit' of github.com:UnosquareBelfast/AdminCore into feature/initial-commit

trojanspike committed:
commit compose edits

linda-mccall committed:
move backend code from root project

trojanspike committed:
ignore idea files - folder

BrownMark reviewed (approved):
Looks good Linda, just a few small issues but we can iron those out later
